### ISSUE TITLE
Speed Factor Issues in SFML Port Branch...

### ISSUE THREAD

**Issue by datahead8888** [Secretchronicles/TSC#477]
In the feature-sfml-port branch, if you load one of the two test levels and jump, you will jump sky high.
Quintus said he believes this may be an issue with the usage of the m_speedfactor variable.  This variable is calculated in scene_manager.cpp in method Play().

The equivalent speed factor code in the devel branch is in framerate.cpp, in method Update().

The jumping code in question in the sfml port branch is in level_player.cpp in method Update_Jump:
https://github.com/Secretchronicles/TSC/blob/feature-sfml-port/tsc/src/level/level_player.cpp#L924-L977

You can see the speed factor and FPS (Frames per second) while playing by pressing Control + D.  This starts the debugging mode.

This task is arguably a high priority, since it will make it difficult to test the SFML port.

**Comment by datahead8888** [comment]
I will look at this a little.  If I end up working on another task, I will unassign myself.

**Comment by datahead8888** [comment]
I've run some tests and drawn some conclusions.  I agree there is an issue with the speed factor:
* The SFML port version uses seconds for elapsed time (m_game_clock.restart().asSeconds()) while devel uses milliseconds (SDL_GetTicks())
* The SFML port version changes the speed factor once each second while the version in devel appears to change it every physics iteration
* There are additional minor differences that appear to be in devel such as logic to check m_force_speed_factor before deciding if the speed factor has changed.  Much of this logic is stripped out, but it isn't as likely to be causing the bug.

I told it to print the relevant variables each time a new FPS was calculated for each second in both the sfml port and devel branches.  There appear to be major discrepancies between the values:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-------------------------------
feature-sfml-port:
Elapsed Time: 0.000253
Speed factor: 0.000272966
FPS Reported: 3664
-------------------------------
-------------------------------
feature-sfml-port:
Elapsed Time: 0.000778
Speed factor: 0.000252004
FPS Reported: 3971
-------------------------------
-------------------------------
feature-sfml-port:
Elapsed Time: 0.000954
Speed factor: 0.000220715
FPS Reported: 4534
-------------------------------
-------------------------------
feature-sfml-port:
Elapsed Time: 0.00052
Speed factor: 0.000411072
FPS Reported: 2433
-------------------------------
===========================================================================
-------------------------------
Devel
Elapsed Time: 16
Speed factor: 0.512
FPS Reported: 59
-------------------------------
-------------------------------
Devel
Elapsed Time: 16
Speed factor: 0.512
FPS Reported: 59
-------------------------------
-------------------------------
Devel
Elapsed Time: 16
Speed factor: 0.512
FPS Reported: 59
-------------------------------
-------------------------------
Devel
Elapsed Time: 17
Speed factor: 0.544
FPS Reported: 59
-------------------------------
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

These are the relevant sections of code (with only my cout's added)::::
Branch feature-sfml-port, file scene_manager.cpp, method Play().  m_speed_factor is only changed once per second:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~c++
        // Measure time we needed for this frame
        m_elapsed_time = m_game_clock.restart().asSeconds();

        // Calculate the framerate, i.e. the amount of frames we can do per second.
        total_elapsed_time += m_elapsed_time;
        if (total_elapsed_time >= 1.0f) {
            m_speedfactor = total_elapsed_time / m_frames_counted;

            //sprintf(fps_text, "FPS: %d Speedfactor: %f", m_frames_counted, m_speedfactor);
            m_framerate_text.setString(fps_text);

            cout << "-------------------------------" << endl;
            cout << "feature-sfml-port:" << endl;
            cout << "Elapsed Time: " << m_elapsed_time << endl;
            cout << "Speed factor: " << m_speedfactor << endl;
            cout << "FPS Reported: " << m_frames_counted << endl;
            cout << "-------------------------------" << endl;

            m_frames_counted = 0;
            total_elapsed_time = 0.0f;


        }
        m_frames_counted++;

        /*
          // Debugging the speedfactor. The below code forces a pause to
          // decrease the speedfactor for debugging purposes.
        struct timespec t;
        t.tv_sec = 0;
        t.tv_nsec = 1000000;
        nanosleep(&t, NULL);
        */
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Branch devel, file framerate.cpp, method Update().  m_speed_factor seems to be changed during each update iteration:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~c++
void cFramerate::Update(void)
{
    const Uint32 current_ticks = SDL_GetTicks();

    // if speed factor is forced
    if (!Is_Float_Equal(m_force_speed_factor, 0.0f)) {
        m_speed_factor = m_force_speed_factor;
        m_elapsed_ticks = static_cast<Uint32>((m_force_speed_factor * 1000) / m_fps_target);

        // change to minimum
        if (m_elapsed_ticks == 0) {
            m_elapsed_ticks = 1;
        }
    }
    // measure speed factor
    else {
        // set elapsed ticks
        m_elapsed_ticks = current_ticks - m_last_ticks;

        // minimum
        if (m_elapsed_ticks == 0) {
            m_elapsed_ticks = 1;
        }
        // maximum
        else if (m_elapsed_ticks > m_max_elapsed_ticks) {
            m_elapsed_ticks = m_max_elapsed_ticks;
        }

        // speed factor calculation for this frame
        m_speed_factor = static_cast<float>(m_elapsed_ticks / (1000 / m_fps_target));
    }

    // speed factor based fps
    m_fps = m_fps_target / m_speed_factor;

    // calculate average fps every second
    if (current_ticks - m_fps_average_framedelay > 1000) {
        m_fps_average = m_frames_counted;

        m_fps_average_framedelay += 1000;
        m_frames_counted = 0;

        std::cout << "-------------------------------" << std::endl;
        std::cout << "Devel" << std::endl;
        std::cout << "Elapsed Time: " << m_elapsed_ticks << std::endl;
        std::cout << "Speed factor: " << m_speed_factor << std::endl;
        std::cout << "FPS Reported: " << m_fps_average << std::endl;
        std::cout << "-------------------------------" << std::endl;
    }
    // count a fps
    else {
        m_frames_counted++;
    }

    // best fps
    if (m_fps > m_fps_best) {
        m_fps_best = m_fps;
    }
    // worst fps
    else if (m_fps < m_fps_worst) {
        m_fps_worst = m_fps;
    }

    m_last_ticks = current_ticks;
}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by datahead8888** [comment]
It seems my first assumption above was incorrect.  There does not appear to be an issue with seconds versus milliseconds -- the code in devel divides the milliseconds for the elapsed ticks by a factor of 1000.

The speed factor algorithm in the feature-sfml-port branch, however is different than in devel.  In addition to calculating the speed factor once each second instead of each physics update iteration, it also removes logic to adjust the speed factor based on the m_fps_target value.  Based on my debugging, m_force_speed_factor had no influence on level play while just jumping, so the removal of this in feature-sfml-port-branch is probably not causing issues.

In the sfml code, I've tried making a hacked version of the code that forces the elapsed time to be hard coded to 15, that accounts for m_fps_target (value of 32), and that thus forces the speedfactor to always be calculated as 0.48:
https://github.com/Secretchronicles/TSC/blob/sfml-port-speedfactor-fix-HACKED/tsc/src/core/scene_manager.cpp#L94

~~~~~~~~~~~~~~~~c++
m_elapsed_time = 15; //FIXED VALUE
.....
m_speedfactor = m_elapsed_time / (1000.0 / 32); //Calculate once per physics iteration.  32 corresonds to the value of m_fps_target in devel
~~~~~~~~~~~~~~~~

By hard coding the elapsed time to 15 in devel as well, the speed factor is always the same in both devel and the feature-sfml-port branch.  Devel then plays fine when I jump in a level, but the sfml branch does not, going way too fast still.  Thus I think there is a second problem other than the speed factor here.

@Quintus, I'm not quite sure what your thought process was when you rewrote the speed factor algorithm or what other changes you made to compensate for the problem.  In IRC you said, "What happens is that the game applies a *way* to strong Y velocit to Alex if you press the Jump key".  If you have any insight, I might be able to look into this bug more; otherwise it would probably be best if you take a look at it again.

**Comment by Quintus** [comment]
> The speed factor algorithm in the feature-sfml-port branch, however is different than in devel. In addition to calculating the speed factor once each second instead of each physics update iteration, it also removes logic to adjust the speed factor based on the m_fps_target value

Since the main loop was one of the very first things to be ported, I think this was before I added my `// OLD` comments everywhere. If there was a removal of code, this was unintentional and I meant to port it later over, but then simply forgot about it (now you see what these `// OLD` comments are good for).

>  Devel then plays fine when I jump in a level, but the sfml branch does not, going way too fast still

There are a *lot* of magic numbers in the TSC code currently, and I remember having fiddled with one of these when I worked on the jumping code so that it now is not the same anymore as it was in OpenGL/SDL TSC. If you check my commits regarding jumping you might find where I changed that number. All these magic numbers in OpenGL/SDL TSC are a hazzle to work with. If you have something like (pseudo example)

~~~~
0.0464 * speedfactor
~~~~

, then what the heck is “0.0464”? You’ll find this everywhere and a dozen times in OpenGL/SDL TSC. Such was a number that I changed when I ported the jumping code. I also de-magicified and changed the gravity system’s values. If you take a look at how `cActor::Update_Gravity` is implemented, you might find the root cause of all of this jumping issue.

> @Quintus, I'm not quite sure what your thought process was when you rewrote the speed factor algorithm

Me neither. Feel free to stash it and redo it. I honestly think I wrote crappy code when I did that.

Vale,
Quintus

**Comment by Quintus** [comment]
I forgot to mention that I extracted the speed factor calculation into its own method now for better overview (and use in TSC’s awful mini mainloops). Apart from that extraction nothing happened to the speedfactor code in `feature-sfml-port`.

Vale,
Quintus

**Comment by datahead8888** [comment]
@Quintus, did you temporarily change any other code to compensate for the speed issues or know of anything else that might be causing the jumping issues?  My tests yesterday seemed to indicate that forcing the speed value to be the same between the feature-sfml-port and devel branches does not fix the issue.  We should probably first get jumping to behave correctly in a temporary test before deciding on how to write the main game loop.

> There are a lot of magic numbers in the TSC code currently

We probably should keep most of the magic numbers for now, until the SFML port is done, since there is a risk of bugs.  This is akin to putting other architectural changes on hold.  We can then start wiping them out in later tasks.  We're going to be cleaning up the code base for a while one way or another.

**Comment by Quintus** [comment]
> did you temporarily change any other code to compensate for the speed issues

No, what I mentioned was all. I just live with it in testing currently.

> anything else that might be causing the jumping issues

It might be a problem with the gravity system as indicated. I have no other ideas, though.

> We should probably first get jumping to behave correctly in a temporary test before deciding on how to write the main game loop.

Okay.

> We probably should keep most of the magic numbers for now, until the SFML port is done, since there is a risk of bugs. This is akin to putting other architectural changes on hold.

Okay.

Vale,
Quintus

**Comment by datahead8888** [comment]
I have written fixes for two issues I've found regarding the speed factor.  They are at https://github.com/Secretchronicles/TSC/commit/a7966fe9a6f315a3cf8bb3d83f0b1aa0ac660f27
I found:
* In actor.cpp in method Update_Position(), it did not account for the current time elapsed (represented through the speed factor) when applying a position change based on velocity.
* In scene_manager.cpp it only calculated the speed factor once every second (rather than once each iteration) and removed the usage of m_fps_target in the calculation.

Jumping seems to work well in the first debugging level.  However, if you play the second debugging test level you will find that Alex now gets stuck inside the ground beneath the starting point.  I can see that major changes were made to how collisions are handled when the position is updated, so it may be related to these changes.

I see two main possibilities:
* The equivalent to cActor::UpdatePosition  in the sfml branch is cSprite::Move in the devel branch.  In this method, it calls Update_Position_Rect() at the end.  This method in turn updates the collision rectangle after movement.  When I fixed the pip issues, one of the bugs was this rectangle not being set properly.  I did not find an equivalent to Update_Position_Rect() in the new sfml code, though it's possible I missed it.
* It could be related to this new comment in cActor::Update_Position(): "Moving first and then checking for collisions will result in the object being inside the wall before the collision detection algorithm actually yields true... Maybe this should be reconsidered"

This is the code for Update_Position_Rect from devel:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~c++
void cSprite::Update_Position_Rect(void)
{
    // if not editor mode
    if (!editor_enabled) {
        m_rect.m_x = m_pos_x;
        m_rect.m_y = m_pos_y;
        // editor rect
        m_start_rect.m_x = m_pos_x;
        m_start_rect.m_y = m_pos_y;
        // collision rect
        m_col_rect.m_x = m_pos_x + m_col_pos.m_x;
        m_col_rect.m_y = m_pos_y + m_col_pos.m_y;
    }
    // editor mode
    else {
        m_rect.m_x = m_start_pos_x;
        m_rect.m_y = m_start_pos_y;
        m_start_rect.m_x = m_start_pos_x;
        m_start_rect.m_y = m_start_pos_y;
        // Do not use m_start_pos_x/m_start_pos_y because col_rect is not the editor/start rect
        m_col_rect.m_x = m_pos_x + m_col_pos.m_x; // todo : startcol_pos ?
        m_col_rect.m_y = m_pos_y + m_col_pos.m_y;
    }

    Update_Valid_Draw();
}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by datahead8888** [comment]
However, I'll need to find the equivalent of m_col_pos.m_x if I rewrite the Update_Position_Rect method.

**Comment by Quintus** [comment]
> I have written fixes for two issues I've found regarding the speed factor.

Very nice, this looks good. The jumping in the minimal levels works fine now. If you edit the XML of `lvl_2.tsclvl` and set a higher Y position for the level player (decrease Y, mind negative value and remember to run `make install` afterwards) you will find it also works fine in that level now. Great work!

However, what did you do with the start menu? It expands/shrinks the menu points now much too quickly. :-)

> This method in turn updates the collision rectangle after movement.

The collision rectangle doesn’t need to be calculated explicitely anymore. This is due to the way the actors are coupled with SFML’s sprites. SFML has an incredibly cool `sf::Transformable` class ([tutorial](http://www.sfml-dev.org/tutorials/2.3/graphics-transform.php)) that `cActor` inherits from. Whenever you now call an SFML transformation method like `move()`, SFML calculates a “transformation” object ([`sf::Transform`](http://www.sfml-dev.org/documentation/2.3/classsf_1_1Transform.php)) from that. You can then use that transform on any rectangle to auto-calculate how the resulting rectangle will look like (when it’s moved, resized, etc.). `cActor` thus only once needs to have the collision rectangle set for position (0|0) (which is done by reading it from the `.settings` file), and when you need the current position and sizing of the collision rectangle (yes, you can resize it together with the actor now!) you can just apply this `sf::Transform` object to the (0|0) collision rectangle. This is what the method [cActor::Get_Transformed_Collision_Rect](https://github.com/Secretchronicles/TSC/blob/0b8b117efc3900773b2b0859bb111738fde26969/tsc/src/objects/actor.cpp#L305-L308) does. The collision rect is thus always automatically correct and you can simply forget about it when moving actors around, the collision system always retrieves the correctly positioned/sized one by calling cActor::Get_Transformed_Collision_Rect().

> "Moving first and then checking for collisions will result in the object being inside the wall before the collision detection algorithm actually yields true... Maybe this should be reconsidered"

This was in old SDL/OpenGL TSC aready, I just added the comment. SDL/OpenGL TSC mitigated the problem by means of an “anti-stuck” concept that made minimal position adjustments if it detected an actor was stuck in a wall. The “anti-stuck” has not yet been ported over to the SFML port.

What I ought to remove is my nonsensical BinaryTree class. I already wrote it when I realised that `std::map` is already implemented as a binary tree as per “The C++ Programming Language”. That is for sure more performant than my own implementation and has a nicer API. That’s only a side note, though.

Vale,
Quintus

**Comment by datahead8888** [comment]
> If you edit the XML of lvl_2.tsclvl and set a higher Y position for the level player (decrease Y, mind negative value and remember to run make install afterwards) you will find it also works fine in that level now.

Yes, after doing this I was able to hop my way through the level.  Jumping still has issues (or hacks if some of the functionality is intended for now), however.

> However, what did you do with the start menu? It expands/shrinks the menu points now much too quickly. 

I made no changes to the start menu.  If, however, you wrote your code to work while we had speed issues, I would expect it to break when the those main speed factor issues were fixed.  Otherwise, this is yet another issue to look into.  By the way, where is the old code for this in the devel branch?

> What I ought to remove is my nonsensical BinaryTree class. I already wrote it when I realised that std::map is already implemented as a binary tree as per “The C++ Programming Language”. That is for sure more performant than my own implementation and has a nicer API. That’s only a side note, though.

As I've said, I think it makes sense to reinvent the wheel sometimes, but in this case, std::map might be a better idea and is something most of the team should already be familiar with.  I have not looked closely as how you are using the tree yet, however.

Per our discussion in IRC, you had said you thought the current collision issues were not due to either of the two items I mentioned in my last post regarding possible collision issue causes.  Thus we cannot merge the speedfactor fixes until we know what is causing Alex to get stuck in the ground.  Once that is fixed, we will have a functional improvement and can merge them.

Playing through the level with the speed factor fixes present, I currently make these observations:
* I can do "multi jumps", jumping again in the air any number of times I want.
* If I hold down the jump button, Alex will multi jump about once and then just climb upwards with a non-stopping jump
* Once moving on the ground, Alex will start sliding and doesn't stop unless the opposite direction is pressed
* Alex can pass through a number of massive objects
* While walking and pressing against a massive object that is not too tall (and is not one of the ones having issues mentioned above), I find that Alex suddenly steps up on top of it.  For a taller pipe, I was able to step into it, standing against the edge.  This could be related to collision issues.
* Running seems to be turned off

My hypothesis is that there are several other bugs to be found, which have been working together to produce unusual behavior.

**Comment by datahead8888** [comment]
Also:
* Falling down a pit does not kill Alex

**Comment by Quintus** [comment]
> I made no changes to the start menu. If, however, you wrote your code to work while we had speed issues, I would expect it to break when those main speed factor issues were fixed

Yes, that’s true. We’re going to have to adapt the menu to the “correct” calculation of the speedfactor now.

> By the way, where is the old code for this in the devel branch?

menu/menu.cpp and menu/menu_data.cpp. Beware the in-memory XML.

> Per our discussion in IRC, you had said you thought the current collision issues were not due to either of the two items I mentioned in my last post regarding possible collision issue causes. Thus we cannot merge the speedfactor fixes until we know what is causing Alex to get stuck in the ground

As said, I will take a look at this, probably this weekend. In the meantime you could fix the timing for the start menu in menu_scene.cpp :-)

> I can do "multi jumps", jumping again in the air any number of times I want.

Jumping code is not ported over completely yet AFAIK, and this is probably the result of that.

> Once moving on the ground, Alex will start sliding and doesn't stop unless the opposite direction is pressed

I find that this sometimes happens, but not always  Usually the auto-stop functionality kicks in and stops him. This is somehow dependant on the game run... That is, you need to restart the game and try again to provoke a different effect. Either auto-stop always works, or never. Something else I will have to look into.

> Alex can pass through a number of massive objects

I can’t reproduce that. Do you have an example?

> While walking and pressing against a massive object that is not too tall (and is not one of the ones having issues mentioned above), I find that Alex suddenly steps up on top of it

I noticed that, and that’s some kind of bug, yes.

> Running seems to be turned off

Not ported over yet.

> Falling down a pit does not kill Alex

Not ported over yet, depends on #461.

> My hypothesis is that there are several other bugs to be found, which have been working together to produce unusual behavior.

That’s likely.

Vale,
Quintus

**Comment by datahead8888** [comment]
> Yes, that’s true. We’re going to have to adapt the menu to the “correct” calculation of the speedfactor now.

Done with commit id 9381b3070e8bc491a753f1bdd61930f79c0bd0b1 in the sfml_speedfactor_fix  branch

> menu/menu.cpp and menu/menu_data.cpp. Beware the in-memory XML

I probably missed it after being tired out after working on the speedfactor fixes some days back.

> Jumping code is not ported over completely yet AFAIK, and this is probably the result of that.

I assume this is part of #434 if code is completely missing; otherwise it should be a new bug ticket.  Perhaps this ticket should be retitled as "speed factor issues".

> I find that this sometimes happens, but not always Usually the auto-stop functionality kicks in and stops him. This is somehow dependant on the game run... That is, you need to restart the game and try again to provoke a different effect. Either auto-stop always works, or never. Something else I will have to look into.

If confirmed as a bug, we might want to open a ticket.  We probably need some bugs filed to help track specific issues.

> I can’t reproduce that. Do you have an example?

I meant the wooden looking blocks toward the start of the second test level.  I guess these may have been passive, however.  The other cases appear to be instances of the other collision bug where he "steps up" onto platforms.

> Running seems to be turned off

> Not ported over yet.

I assume this is part of #434.

**Comment by Quintus** [comment]
> Done with commit id 9381b30 in the sfml_speedfactor_fix branch

Thanks. I just tested it, works like a charme. :-)

> I assume this is part of #434 if code is completely missing; otherwise it should be a new bug ticket. 

Let me verify against the code before filing a new ticket.

> If confirmed as a bug, we might want to open a ticket. We probably need some bugs filed to help track specific issues.

Agreed. Feel free to file one.

>  meant the wooden looking blocks toward the start of the second test level.

They’re set to “half-massive”, not “massive”. Jumping through them is allowed. However, what does not work yet (because not yet ported) is standing on a half-massive object.

Vale,
Quintus

**Comment by Quintus** [comment]
> Let me verify against the code before filing a new ticket.

Okay I have a new finding there. The issue directly correlates with the sliding problem. If you have a game run that contains the side-sliding problem, you also can do double jumps in the air. If however you have a game run that doesn’t contain the side-sliding problem, you also can’t do double jumps in the air. There is an issue with the keyboard handling I infer from that (maybe keydown/keyup event not received or so).

Vale,
Quintus
